libr/include/r_cons.h

```diff
@@ -230,6 +231,7 @@ typedef struct r_cons_canvas_t {
 	char *b;
 	int blen;
 	const char * attr;//The current attr (inserted on each write)
+	int attr_dyn_alloc;//if attr atttribute is dynamically allocated, it needs to be free'd
```

Comment:
this should be bool.. but we really need to differentiate that? its easier to just only handle dynamically allocated attributes instead of duping the logic

Reply:
I added flags since there is no way to identify whether the address pointed by `attr` is dynamically allocated or not. For example, in [set_attr](https://github.com/radare/radare2/blob/master/libr/cons/canvas.c#L221) function, attr is value is set to dynamically allocated address while at few other places like [this](https://github.com/radare/radare2/blob/master/libr/cons/canvas.c#L397), attr attribute is set to a static value. And to fix this thing, I had to add a flag.

Similarly when the attr value is copied in [stamp_attr](https://github.com/radare/radare2/blob/master/libr/cons/canvas.c#L196) function, we don't have a way of saying whether the copied value points to dynamically allocated string or to static string. Thus when freeing or clearing the `RConsCanvas`, the other flag in r_cons_canvas_attr_t help in deciding whether to free value of `a` or not.

The only reason I am doing strdup everywhere is because I didn't know if the value of `c->attr` is copied multiple times in attrs array or just one time. I can check on that and make changes accordingly.

Reply:
Please let me know if you have some other way to fix this. I know this is a bad way to fix.

Reply:
I also noticed when I was fixing some mem leaks in canvas.c b/c this one was hard so i left out.

This time I would agree with pancake (I am often a dissenter). `strdup` every attribute including constants such as Color_RESET Color_RED may be easier. `RConsCanvas::attrslen` is cleared when repainting and is bound by the number of different displayed objects (usually <100). `strdup` waste should be acceptable.

BTW, the binary search implemented in 
`static const char **attr_at(RConsCanvas *c, int loc) {`
is incorrect.